contracts/RockPaperScissors.sol

```diff
@@ -0,0 +1,286 @@
+        uint256 value
+    );
+
+    event PlayerTwoMoves(
+        address indexed player,
+        bytes32 indexed gameKey,
+        uint256 value
+    );
+
+    event PlayerTwoReveals(
+        address indexed player,
+        bytes32 indexed gameKey,
+        uint8 move
+    );
+
+    event WinnerRewarded(
+        address indexed player,
+        bytes32 indexed gameKey,
+        uint256 amount
+    );
+
+    event  ForfeitPaid(
+        address indexed player,
+        bytes32 indexed gameKey,
+        uint256 amount
+    );
+
+    event WithDraw(
+        address indexed withdrawer,
+        uint amount
+    );
+
+    // The creator's move is embedded in the game key
+    function playerOneMove(bytes32 gameKey, uint stake) external payable whenNotPaused {
+
+        require(gameKey != bytes32(0), "Game key cannot be empty");
+        require(stake > 0, "Stake cannot be zero");
+
+        Game storage game = games[gameKey];
+        require(game.step == Steps.INIT, INVALID_STEP_MSG);
+
+        game.playerOne = msg.sender;
+        game.stake = stake;
+        game.step = Steps.PLAYER_ONE_MOVE;
+        depositStake(stake);
+
+        emit PlayerOneMoves(msg.sender, gameKey, stake);
+    }
+
+    /*
+    * The opponent joins the game
+    */
+    function playerTwoMove(bytes32 gameKey, bytes32 moveHash) external payable whenNotPaused {
+
+        Game storage game = games[gameKey];
+        uint256 stake = game.stake;
+        require(stake != 0, GAME_NOT_FOUND_MSG);
+        require(game.step == Steps.PLAYER_ONE_MOVE, INVALID_STEP_MSG);
+
+        game.playerTwo = msg.sender;
+        game.playerTwoMoveHash = moveHash;
+        game.step = steps.PLAYER_TWO_MOVE;
+        depositStake(stake);
+
+        emit PlayerTwoMoves(msg.sender, msg.value, gameKey);
+    }
+
+    function depositStake(uint256 stake) internal {
+        uint256 senderBalance = balances[msg.sender];
+
+        if (msg.value == stake) {
+            return;
+        }
+
+        if (msg.value < stake) {
+            require(senderBalance >= stake.sub(msg.value), "Insufficient balance");
+            balances[msg.sender] = senderBalance.sub(stake.sub(msg.value));
+        } else {
+            balances[msg.sender] = senderBalance.add(msg.value.sub(stake));
+        }
+
+        Deposit(msg.sender, msg.value);
+    }
+
+    function playerOneReveal(bytes32 gameKey, bytes32 secret, uint8 move) external whenNotPaused {
+
+        require(move < 3, INVALID_MOVE_MSG);
+        Game storage game = games[gameKey];
+        uint256 stake = game.stake;
+        require(stake != 0, GAME_NOT_FOUND_MSG);
+        require(game.step == Steps.PLAYER_TWO_MOVE, INVALID_STEP_MSG);
+        address playerOne = game.playerOne;
+        require(msg.sender == playerOne, ONLY_PLAYER_ONE_CALL_MSG);
+
+
+        //Validate move
+        bytes32 expectedMoveHash = createGameKey(secret, move);
+        require(gameKey == expectedMoveHash, "Move and secret do not match");
```

Comment:
Lol. What would you think of this fictitious function of mine?

```solidity
function add(uint left, uint right, uint sum) returns (uint) {
    require(left + right == sum);
    return sum;
}
```

Reply:
I hope I understand your explanation although I am thinking that the following make it different:

1. We have verified that 'gameKey' must be the same as that used for the game in movePlayerOne():

```
        Game storage game = games[gameKey];
        require(game.step == uint8(Steps.PLAYER_TWO_MOVE), INVALID_STEP_MSG);
        require(msg.sender == playerOne, INVALID_PLAYER_MSG);
```
2. The hashing also incorporates the msg.sender

Reply:
My point here is that why do you do `require(gameKey == expectedMoveHash` where `gameKey` is passed in the function and `expectedMoveHash` is calculated in the contract? You correctly identified that `expectedMoveHash` is the stronger one and `gameKey` is the weak one. So why not skip the `==` altogether and use `expectedMoveHash` wherever there is a need for `gameKey`?

Reply:
Ah yes. I see. Took me a while to make the conceptual mental leap.

I think my initial reasoning was to provide an explicit user friendly precise error but given I already have a `require()` that already checks the correct step, that would be sufficient to implicitly indicate the `secret` and `move` are incorrect.